### ISSUE TITLE
Can I train model, in which one image is custom and another image is on which predefined model is trained on.

### ISSUE THREAD

**Issue by pranjalsharma26** [sglvladi/TensorFlowObjectDetectionTutorial#29]
By this technique I need to classify two objects, first is my own custom image and another is image say car on which model is trained on as I took some predefined model trained on COCO dataset. So this task is possible ? 
Please reply asap.
Thanks in advance.

**Comment by sglvladi** [comment]
@pranjalsharma26 I am not very clear as to what you mean. Could you please elaborate more on your explanation?

**Comment by pranjalsharma26** [comment]
@sglvladi I want to detect license plate with this technique and car at a time. Is this possible that I need to make a license plate as my custom object and it will detect car by its own when I choose a pre-trained model say ssd_mobilenet_v2_COCO.

**Comment by sglvladi** [comment]
@pranjalsharma26 I realise that this a really delayed response, but I haven't really been able to respond. Hopefully my answer can prove useful for you or anyone else who might come across this issue in the future.

My understanding is that it is not possible to retrain a model so as to "add" an extra object to an existing network. If you really want to reuse an existing model, the best course of action would be to actually have two models. Assuming you want to use ``ssd_mobilenet_v2_coco``, this is what I would propose:

1. Follow the tutorial to train a ``ssd_modilenet_v2_coco`` with a custom dataset of number plates. Once the training is done and you are happy with your results, the resulting model will be your number plate detector.
2. Use a copy of the ``ssd_modilenet_v2_coco`` model you used as your training checkpoint for the number plate detector, as your car detector.

Below is a sample code that should allow you to combine the two models in a single script:

``` python

import cv2
import numpy as np
import tensorflow as tf

from utils import label_map_util as lm_util
from utils import visualization_utils as vis_util

config = tf.ConfigProto(
#    Uncomment the lines below if you're using a GPU
#    gpu_options = tf.GPUOptions(per_process_gpu_memory_fraction=0.3),
#    device_count = {'GPU': 1}
)

# ----------------------- Number Plate Model --------------------------------- #
PLATE_MODEL_PATH = "<PATH_TO_PLATE_MODEL_.pb_FILE>"
PLATE_LABEL_PATH = "<PATH_TO_PLATE_LABEL_.pbtxt_FILE>"
plate_detection_graph = tf.Graph()
with plate_detection_graph.as_default():
    plate_od_graph_def = tf.GraphDef()
    with tf.gfile.GFile(str(PLATE_MODEL_PATH), 'rb') as fid:
        serialized_graph = fid.read()
        plate_od_graph_def.ParseFromString(serialized_graph)
        tf.import_graph_def(plate_od_graph_def, name='')
    plate_session = tf.Session(graph=plate_detection_graph, config=config)
plate_category_index = lm_util.create_category_index_from_labelmap(
    PLATE_LABEL_PATH,
    use_display_name=True)
# ---------------------------------------------------------------------------- #


# --------------------------- Car Model -------------------------------------- #
CAR_MODEL_PATH = "<PATH_TO_CAR_MODEL_.pb_FILE>"
CAR_LABEL_PATH = "<PATH_TO_CAR_LABEL_.pbtxt_FILE>"
car_detection_graph = tf.Graph()
with car_detection_graph.as_default():
    car_od_graph_def = tf.GraphDef()
    with tf.gfile.GFile(str(CAR_MODEL_PATH), 'rb') as fid:
        serialized_graph = fid.read()
        car_od_graph_def.ParseFromString(serialized_graph)
        tf.import_graph_def(car_od_graph_def, name='')
    car_session = tf.Session(graph=car_detection_graph, config=config)
car_category_index = lm_util.create_category_index_from_labelmap(
    PLATE_LABEL_PATH,
    use_display_name=True)
# ---------------------------------------------------------------------------- #


# ----------------------- Detection Functions -------------------------------- #
def detect_plates(image):
    """ This detects plates in an image and returns the results """

    # Extract image tensor
    image_tensor = plate_detection_graph.get_tensor_by_name('image_tensor:0')
    # Extract detection boxes
    boxes = plate_detection_graph.get_tensor_by_name('detection_boxes:0')
    # Extract detection scores
    scores = plate_detection_graph.get_tensor_by_name('detection_scores:0')
    # Extract detection classes
    classes = plate_detection_graph.get_tensor_by_name('detection_classes:0')
    # Extract number of detections
    num_detections = plate_detection_graph.get_tensor_by_name('num_detections:0')
    # Actual detection.
    (boxes, scores, classes, num_detections) = plate_session.run(
        [boxes, scores, classes, num_detections],
        feed_dict={image_tensor: image})
    return boxes, scores, classes, num_detections


def detect_cars(image):
    """ This detects cars in an image and returns the results """

    # Extract image tensor
    image_tensor = car_detection_graph.get_tensor_by_name('image_tensor:0')
    # Extract detection boxes
    boxes = car_detection_graph.get_tensor_by_name('detection_boxes:0')
    # Extract detection scores
    scores = car_detection_graph.get_tensor_by_name('detection_scores:0')
    # Extract detection classes
    classes = car_detection_graph.get_tensor_by_name('detection_classes:0')
    # Extract number of detections
    num_detections = car_detection_graph.get_tensor_by_name('num_detections:0')
    # Actual detection.
    (boxes, scores, classes, num_detections) = car_session.run(
        [boxes, scores, classes, num_detections],
        feed_dict={image_tensor: image})
    return boxes, scores, classes, num_detections
# ---------------------------------------------------------------------------- #


# Define the video stream
cap = cv2.VideoCapture(0)

# Merge category indices
category_index = {**plate_category_index, **car_category_index}

# Detection
while True:

    # Read frame from camera
    ret, image_np = cap.read()
    # Expand dimensions since the model expects images to have shape: [1, None, None, 3]
    image_np_expanded = np.expand_dims(image_np, axis=0)

    # Detect plates
    (plate_boxes, plate_scores, plate_classes, plate_num_detections) = detect_plates(image_np_expanded)

    # Detect cars
    (car_boxes, car_scores, car_classes, car_num_detections) = detect_cars(image_np_expanded)

    # Merge the detections
    boxes = np.concatenate((plate_boxes, car_boxes), 1)
    scores = np.concatenate((plate_scores, car_scores), 1)
    classes = np.concatenate((plate_classes, car_classes), 1)
    num_detections = plate_num_detections + car_num_detections

    # Visualization of the results
    vis_util.visualize_boxes_and_labels_on_image_array(
        image_np,
        np.squeeze(boxes),
        np.squeeze(classes).astype(np.int32),
        np.squeeze(scores),
        category_index,
        use_normalized_coordinates=True,
        line_thickness=8)

    # Display output
    cv2.imshow('object detection', cv2.resize(image_np, (800, 600)))

    if cv2.waitKey(25) & 0xFF == ord('q'):
        cv2.destroyAllWindows()
        break
```

It seems to work, however here are some notes:
- It will be slower that running a single model, as each frame will be run through two models.
- Notice the commented out lines at the beginning of the script relating to the GPU. Tensorflow seems to like to eat up all of the GPU VRAM for each session. Therefore, to ensure you can run 2 sessions, the lines in the config limit the amount of available to VRAM to each session, otherwise the script will crash and you will get ``Could not create cudnn handle: CUDNN_STATUS_ALLOC_FAILED`` errors.
- When setting up the label map for the ``plate`` model prior to training (see [here](https://tensorflow-object-detection-api-tutorial.readthedocs.io/en/latest/training.html#creating-label-map)) make sure to give a high value to the ``id`` of the label, so that there is not chance it will coincide with the ``id`` of the ``car`` label in your chosen model. For example I would do something like following, just to be safe :
```
item {
    id: 107995
    name: 'plate'
}
``` 
- If you use a ready made model (e.g. ``ssd_modilenet_v2_coco``), you will most probably have to do some form of post-processing (possible inside ``detect_cars()`` function) to ensure that only car detections are returned, as you will most certainly get other objects as well.

Hope this helps.
